sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,378 @@
+
+This document proposes the specialization of `sycl::marray` to add support for
+storing complex numbers in arrays.
+
+The proposal includes overloading the existing math functions to support complex
+marrays and adding member functions to simplify accessing, setting marray
+values, and constructing complex marrays.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_COMPLEX_MARRAY` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== Marray Complex Class Specialization
+
+The user interface of the `sycl::marray<sycl::ext::oneapi::complex, N>` class is
+changed compared to the SYCL-2020 generic `sycl::marray` interface.
+
+This proposition changes the marray interface when specialized for the
+`sycl::ext::oneapi::complex` type.
+
+The marray complex specialization is trivially copyable, and the type trait
+`is_device_copyable` should resolve to `std::true_type`.
+
+The marray definition used within this proposal assumes that any operator the
+`sycl::marray` class defines is only implemented if the marray's value type also
+implements the operator.
+
+For example, `sycl::marray<sycl::ext::oneapi::complex, N>` does not implement
+the modulus operator as `sycl::ext::oneapi::complex` does not support it.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+// Specialization of exiting `marray` class for `sycl::ext::oneapi::complex`
+template <typename T, std::size_t NumElements>
+class marray<sycl::ext::oneapi::complex<T>, NumElements> {
+private:
+  using ComplexDataT = sycl::ext::oneapi::complex<T>;
+
+public:
+  using value_type = ComplexDataT;
+  using reference = ComplexDataT &;
+  using const_reference = const ComplexDataT &;
+  using iterator = ComplexDataT *;
+  using const_iterator = const ComplexDataT *;
+
+public:
+  constexpr marray();
+
+  explicit constexpr marray(const ComplexDataT &arg);
+
+  template <typename... ArgTN>
+  constexpr marray(const ArgTN &... args);
+
+  constexpr marray(const marray<ComplexDataT, NumElements> &rhs);
+  constexpr marray(marray<ComplexDataT, NumElements> &&rhs);
+
+  // Available only when: NumElements == 1
+  template <typename = typename std::enable_if<NumElements == 1>>
+  operator ComplexDataT() const;
+
+  static constexpr std::size_t size() noexcept;
+
+  // real and imag
+  marray<T, NumElements> real() const;
+  marray<T, NumElements> imag() const;
+
+  // subscript operator
+  reference operator[](std::size_t i);
```

Comment:
This requires that the internal data representation is an array of complex. Are we reasonable sure no implementation wants the data storage to be one array for real and one for imag?

Reply:
Please, let me know if I misunderstood your feedback, but that's what the sycl::marray specialization for sycl::complex is about.
If needed, a user could use the `real()` or `imag()` method to let him work on the data split.

@TApplencourt, what do you think?

Reply:
There is already an implementation of this complex specialization of marray within this repo https://github.com/argonne-lcf/SyclCPLX.
No issues have been raised about having an array of complex rather than two arrays for `real` and `imag`.

Also, this PR is for the proposed directory and later will be moved onto the experimental.
We'll not be tied to this proposal implementation.

I believe it is conceivable to have an array of complex for this proposal as a start, and if needed we'll change the structure layout.